### ISSUE TITLE
ADB reverse on a USB device

### ISSUE THREAD

**Issue by SyedAhkam** [cocool97/adb_client#63]
Hi @cocool97,

I started a new side project where I need to execute adb reverse while maintaining a physical connection with the device over USB although It seems right now only the server equivalent has forward & reverse commands implemented. Mind taking a look at it?

**Comment by cocool97** [comment]
I'll have a look at it yes :)

**Comment by cli-s1n** [comment]
Feature has been implemented in #73, can you try it ? :)

**Comment by cli-s1n** [comment]
`forward` is still missing, work in progress !

**Comment by SyedAhkam** [comment]
Thanks @cli-s1n !!

I'll try it soon and report back here.

**Comment by SyedAhkam** [comment]
Hi, I was trying this out but running into this error:

```rs
pub fn reverse_port_adb() -> adb_client::Result<()> {
    let mut device = ADBUSBDevice::autodetect()?;

    device.connect()?;
    device.reverse(ADBProtoPort::TCP(TCP_PORT), ADBProtoPort::TCP(TCP_PORT))?;

    Ok(())
}
```
```
thread 'main' panicked at src/main.rs:17:29:
failed to reverse port with adb: UsbError(Access)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
I assume this has to do with on-device authorization but I thought the call to connect will handle the prompt for authorization as in regular ADB. I also tried to give authorization / access to device using regular ADB then proceeded to use `adb_client` but it still errors out. Am I doing something stupid?

**Comment by cocool97** [comment]
You don't need to call `connect()` yourself (method should be internal to crate).

An example is provided in `adb_cli`, you can have a look at it. 

Can you try running adb_cli with `usb reverse PORT PORT` ? Is it working ? 

ADB server must also be killed as it takes the USB handle on the device

**Comment by SyedAhkam** [comment]
```
❯ cargo run --release -- usb shell
    Finished `release` profile [optimized] target(s) in 0.26s
     Running `/home/syed/dev/OSS/adb_client/target/release/adb_cli usb shell`
Error: Access denied (insufficient permissions)
```

Nope, even usb shell doesn't work. Tried with multiple devices and USB cables. 

Authorization prompt appears though. Made sure to `sudo adb kill-server` before too.

**Comment by cocool97** [comment]
Already got this issue. I've deduced this is some related to the wrong state of adbd process on the device.

To fix I just activate / deactivate USB debugging and reset permissions to make it work

**Comment by SyedAhkam** [comment]
I can't get it to work still using your workaround. Can we open a separate issue to track this?

**Comment by cocool97** [comment]
Yes you can, of course ! Can you just confirm you have the same issue on the master branch as well ?
